### PR TITLE
Handle attestations with missing block 

### PR DESCRIPTION
**Pull request by terencechain**
This PR handles incoming attestations that voted for a block but the block is not in DB. It saves the pending attestations and requests for the missing block. The saved pending attestations gets deleted if it falls an epoch behind. 

I also refactored `validateAggregatedAtt` into its own function in  `validate_aggregate_proof.go`, that way we can reuse them to validate pending attestation.

To test it, check the run time logs for the following pattern:
```go
[2020-02-01 10:43:11]  INFO sync: Requesting block for pending attestation blockRoot=c6fc9679c9c3
[2020-02-01 10:43:11]  INFO sync: Processed pending block and cleared pending block in cache blockRoot=c6fc9679c9c3 slot=164014
[2020-02-01 10:43:17]  INFO sync: Verified and saved pending attestations to pool blockRoot=c6fc9679c9c3 pendingAttsCount=827
```

Check the metrics:
```go
# Count the number of times attestation recovered because of missing block
beacon_attestations_recovered_total 3943
# Count the number of times a missing block recovered from attestation vote.
beacon_blocks_recovered_from_attestation_total 8
```

### CODE REVIEW

**Comment by prestonvanloon** [comment]
Please take a look at the prometheus metric naming conventions doc. 

https://github.com/prometheus/docs/blob/master/content/docs/practices/naming.md

I recommend to change the names to use the _total suffix and beacon_ prefix.
- `beacon_attestations_recovered_total`
- `beacon_blocks_recovered_from_attestation_total`

File: beacon-chain/sync/validate_aggregate_proof.go [hunk 1]
```diff
@@ -61,8 +59,25 @@ func (r *Service) validateAggregateAndProof(ctx context.Context, pid peer.ID, ms
...
+	if !r.validateAggregatedAtt(ctx, m) {
+		return false
+	}
+
+	msg.ValidatorData = m
+
+	return true
+}
+
+func (r *Service) validateAggregatedAtt(ctx context.Context, a *ethpb.AggregateAttestationAndProof) bool {
```
Comment (prestonvanloon) [inline beacon-chain/sync/validate_aggregate_proof.go]: Why refactor this into another method? `validateAggregateAndProof` describes the behavior of `validateAggregatedAtt` or are they different?

**Comment by terencechain** [comment]
> Please take a look at the prometheus metric naming conventions doc.
> 
> https://github.com/prometheus/docs/blob/master/content/docs/practices/naming.md
> 
> I recommend to change the names to use the _total suffix and beacon_ prefix.
> 
> * `beacon_attestations_recovered_total`
> * `beacon_blocks_recovered_from_attestation_total`

Got it. Thanks

File: beacon-chain/sync/pending_attestations_queue.go [hunk 2]
```diff
@@ -0,0 +1,160 @@
...
+		// a node will remove the key from the map to avoid dangling keys.
+		if len(s.blkRootToPendingAtts[bRoot]) == 0 {
+			delete(s.blkRootToPendingAtts, bRoot)
+			numberOfBlocksNotRecoveredFromAtt.Inc()
+		}
+	}
+}
+
+// This returns the current slot based on the time now.
+func (s *Service) currentSlot() uint64 {
```
Comment (prestonvanloon) [inline beacon-chain/sync/pending_attestations_queue.go]: Can you reuse this? https://github.com/prysmaticlabs/prysm/blob/c2fbb40909d7bb2a2af493aa8fd1f125c1d0bfde/beacon-chain/blockchain/process_block_helpers.go#L265

File: beacon-chain/sync/validate_aggregate_proof.go (hunk 1 above)
Comment (terencechain) [reply beacon-chain/sync/validate_aggregate_proof.go]: I want to re-use second half of `validateAggregateAndProof` to validate pending attestations when the missing block arrives. And I dont really want to use `validateAggregateAndProof` to validate pending attestation, it takes in `pid peer.ID, msg *pubsub.Message` which makes it hard to satisfy (pending attestation can also be unaggregated)


With the new refactor:

```go
func validateAggregateAndProof() {

check if it is syncing

decodePubsubMessage

check if the attestation is already in the pool

validateAggregatedAtt // This is reused for processing pending attestation

}
```

File: beacon-chain/sync/pending_attestations_queue.go (hunk 2 above)
Comment (terencechain) [reply beacon-chain/sync/pending_attestations_queue.go]: This depends on the service struct: `s.genesisTime` 

Do you think it's a better pattern if I export `CurrentSlot()` from block chain service and use it in other services?

Comment (prestonvanloon) [reply beacon-chain/sync/pending_attestations_queue.go]: Yeah, you are using s.genesisTime already with the `s.chain.GenesisTime()`

https://github.com/prysmaticlabs/prysm/blob/0d58a463a1709ff9f074573453e99e3e9a89fa23/beacon-chain/blockchain/chain_info.go#L181-L183

File: beacon-chain/sync/validate_aggregate_proof.go (hunk 1 above)
Comment (prestonvanloon) [reply beacon-chain/sync/validate_aggregate_proof.go]: Sounds good thanks

File: beacon-chain/sync/pending_attestations_queue.go (hunk 2 above)
Comment (terencechain) [reply beacon-chain/sync/pending_attestations_queue.go]: done

File: beacon-chain/sync/pending_attestations_queue.go [hunk 3]
```diff
@@ -0,0 +1,163 @@
...
+						}
+						numberOfAttsRecovered.Inc()
+
+						// Broadcasting the attestation again once a node is able to process it.
+						if err := s.p2p.Broadcast(ctx, att); err != nil {
+							log.WithError(err).Error("Failed to broadcast")
+						}
+					}
+				} else {
+					fmt.Println("aggregated")
```
Comment (nisdas) [inline beacon-chain/sync/pending_attestations_queue.go]: remove this

File: beacon-chain/sync/pending_attestations_queue.go [hunk 4]
```diff
@@ -0,0 +1,163 @@
...
+	// the attestations in the queue are still valid. Attestations will
+	// be deleted from the queue if invalid (ie. getting staled from falling too many slots behind).
+	s.validatePendingAtts(ctx, s.chain.CurrentSlot())
+
+	for bRoot, attestations := range s.blkRootToPendingAtts {
+		// Has the pending attestation's missing block arrived yet?
+		if s.db.HasBlock(ctx, bRoot) {
+			numberOfBlocksRecoveredFromAtt.Inc()
+			for _, att := range attestations {
+				// The pending attestations can arrive in both aggregated and unaggregated forms,
```
Comment (nisdas) [inline beacon-chain/sync/pending_attestations_queue.go]: this whole following chunk has multiple conditionals and seems to be doing a lot. Can it be abstracted into smaller methods ?

File: beacon-chain/sync/pending_attestations_queue.go [hunk 5]
```diff
@@ -0,0 +1,161 @@
...
+// root of the missing block. The value is the list of pending attestations
+// that voted for that block root.
+func (s *Service) savePendingAtt(att *ethpb.AggregateAttestationAndProof) {
+	s.pendingAttsLock.Lock()
+	defer s.pendingAttsLock.Unlock()
+
+	root := bytesutil.ToBytes32(att.Aggregate.Data.BeaconBlockRoot)
+
+	_, ok := s.blkRootToPendingAtts[root]
+	if !ok {
```
Comment (nisdas) [inline beacon-chain/sync/pending_attestations_queue.go]: you could just do
```suggestion
		if !ok {
		s.blkRootToPendingAtts[root] = []*ethpb.AggregateAttestationAndProof{att}
		return
	} 
	s.blkRootToPendingAtts[root] = append(s.blkRootToPendingAtts[root], att)
```

File: beacon-chain/sync/pending_attestations_queue.go (hunk 3 above)
Comment (terencechain) [reply beacon-chain/sync/pending_attestations_queue.go]: Done

File: beacon-chain/sync/pending_attestations_queue.go (hunk 4 above)
Comment (terencechain) [reply beacon-chain/sync/pending_attestations_queue.go]: It's actually not bad, it's just comments that seem long. I prefer to keep it as it is for clear commentary.  Each if else is just 3-4 lines long

File: beacon-chain/sync/pending_attestations_queue.go (hunk 5 above)
Comment (terencechain) [reply beacon-chain/sync/pending_attestations_queue.go]: Thanks!

**Pull request merged**